### DISCUSSION TITLE
Leds

### DISCUSSION THREAD

**Discussion by kilehh** [tttapa/Control-Surface#357] [Q&A] [answered]
Hey!

I want to connect 32 leds to arduino mega, i have no clue how could connect them. I dont have shift registers for them and i cannot buy them now, i only have 2 multiplexers CD74HC4067. Can i connect leds to arduino with multiplexers using your library? Also could i make a led matrix with this library. Could there be another solution?

Thanks!

**Comment by tttapa** [comment, marked as answer]
No, you can't use a multiplexer to drive LEDs. A CD74HC4067 just consists of 16 switches that connect between one common pin and 16 “channel” pins. Only one of the switches can be connected at any given instant. This means that if you try to drive LEDs using a multiplexer, you can only turn on one of them at a time, which is probably not what you want. A workaround is to scan through all outputs really quickly to create a persistence of vision, but this requires you to spend time updating the LEDs all the time, eating up precious CPU time on your microcontroller. The common solution is to use a special-purpose chip to do the scanning, so you don't have to worry about it in your microcontroller code. The same applies to matrices. You _could_ do this on a microcontroller, but nobody actually does this because using the right driver chip in the first place is much easier.

Control Surface doesn't support multiplexing LEDs or scanning LED matrices directly, but it does support LED drivers and MAX7219 LED matrix drivers.

If you have an Arduino Mega, you should be able to drive all 32 LEDs from IO pins directly, just make sure you don't go past the ATmega2560 current limits.

**Reply by kilehh** [reply]
I can reduce the led count down to 24 but i have read that mega2560 can support only 200mA which means i could drive only 10 leds. Maybe i could use external power supply if so how? I have seen people power externaly like one led but multiple?

**Reply by tttapa** [reply]
> mega2560 can support only 200mA which means i could drive only 10 leds.

You usually don't run your LEDs at 20mA. 5mA is plenty for a normal indicator LED.

> Maybe i could use external power supply if so how?

If you want to drive multiple LEDs that draw more current than your microcontroller can provide, the only option is to use external LED drivers, there's no way around that. If the total current consumption is less than ~400mA or so, you can power these LED drivers from the 5V line of the Arduino. If the total current is larger, you'll need an external power supply.

Either way, you'll need some kind of driver, be it discrete transistors or LED driver ICs. I'd recommend the latter.

**Reply by kilehh** [reply]
I will try to run them at 5mA but i am ordering a LED driver as soon as i can. 

Big thank you for the help!
